src/middlewared/middlewared/plugins/service.py

```diff
@@ -807,14 +807,34 @@ def _initializing_smartd_pid(self):
             asyncio.ensure_future(self.restart('collectd'))
 
     async def _restart_ups(self, **kwargs):
+        await self.middleware.call('service._restart_ups_job')
+
+    @job(lock='restart_upsmon')
+    async def _restart_ups_job(self, job, **kwargs):
         await self.middleware.call('ups.dismiss_alerts')
         await self.middleware.call('etc.generate', 'ups')
-        await self._service("nut", "stop", force=True, **kwargs)
-        await self._service("nut_upsmon", "stop", force=True, **kwargs)
-        await self._service("nut_upslog", "stop", force=True, **kwargs)
-        await self._service("nut", "restart", **kwargs)
-        await self._service("nut_upsmon", "restart", **kwargs)
-        await self._service("nut_upslog", "restart", **kwargs)
+        await self._service("nut", "stop", force=True, onetime=True)
+        # We need to wait on upsmon service to die properly as multiple processes are
+        # associated with it and in most cases they haven't exited when a restart is initiated
+        # for upsmon which fails as the older process is still running.
+        # Also we retrieve upsmon processes before executing a stop in case any process
+        # becomes a zombie and psutil fails to retrieve the name for it
+        upsmon_processes = set()
+        for proc in psutil.process_iter():
```

Comment:
I am not comfortable using this in the event loop. How about `pgrep`?

Reply:
Sure, i can have it updated asap